packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -781,7 +788,51 @@ function use<T>(usable: Usable<T>): T {
 }
 
 function useMemoCache(size: number): Array<any> {
-  throw new Error('Not implemented.');
+  let memoCache = null;
+  // Fast-path, load memo cache from wip fiber if already prepared
+  let updateQueue: FunctionComponentUpdateQueue | null = (currentlyRenderingFiber.updateQueue: any);
+  if (updateQueue !== null) {
+    memoCache = updateQueue.memoCache;
+  }
+  // Otherwise clone from the current fiber
+  // TODO: not sure how to access the current fiber here other than going through
+  // currentlyRenderingFiber.alternate
+  if (memoCache === null) {
+    const current: Fiber | null = currentlyRenderingFiber.alternate;
+    if (current !== null) {
+      const currentUpdateQueue: FunctionComponentUpdateQueue | null = (current.updateQueue: any);
+      if (currentUpdateQueue !== null) {
+        const currentMemoCache: MemoCache | null = currentUpdateQueue.memoCache;
+        if (currentMemoCache !== null) {
+          memoCache = {
+            data: currentMemoCache.data.map(array => array.slice()),
+            index: 0,
+          };
+        }
+      }
+    }
+  }
+  // Finally fall back to allocating a fresh instance
+  if (memoCache === null) {
+    memoCache = {
+      data: [],
+      index: 0,
+    };
+  }
+  if (updateQueue === null) {
+    updateQueue = createFunctionComponentUpdateQueue();
+    currentlyRenderingFiber.updateQueue = updateQueue;
+  }
+  updateQueue.memoCache = memoCache;
```

Comment:
this whole thing is so tedious without nullish coalescing. is there a reason we don't use that?

```javascript
let memoCache = currentlyRenderFiber.updateQueue ?? null;
if (memoCache === null) {
  const currentMemoCache = currentlyRenderingFiber.alternate?.updateQueue?.memoCache ?? null;
  if (currentMemoCache !== null) {
    memoCache = cow(currentMemoCache);
  }
}
...and similar
```

Reply:
Mostly that we're paranoid about compiler output and it encourages people to reduce redundant type checks by combining branches

Reply:
It's good if code that becomes long after compilation is long in the source. So that we see when we're blowing up the code size. We generally avoid features that generate disproportionate bundler output.

Reply:
Fair, though I wonder if there are other ways to achieve those ends. We have a bot that alerts us to unexpected code size increases, and the type system (if used more) would help us with data modeling. A lot of types are `any` or `mixed` right now and so you have to code more defensively than may be necessary.